stgit/commands/common.py

```diff
@@ -561,11 +569,14 @@ def update_commit_data(cd, options):
             add_sign_line(cd.message, sign_str,
                           cd.committer.name, cd.committer.email))
 
-    # Let user edit the commit message manually, unless
-    # --save-template or --message was specified.
+    # Edit the commit message manually if:
+    # - A message was specified, but --edit was also specified.
+    # - No message was specified, and --save_template was also not specified
     if (
-        not getattr(options, 'save_template', None)
-        and getattr(options, 'message', None) is None
+        (getattr(cd, 'message', None)
+         and getattr(options, 'edit', None))
+        or (not getattr(cd, 'message', None)
+            and not getattr(options, 'save_template', None))
```

Comment:
Using this `update_commit_data()` function in `imprt.py` is certainly proving to be problematic.

Consider the following alternative predicate:

```
if getattr(options, 'edit', None) or (
    not getattr(options, 'save_template', False)
    and getattr(options, 'message', None) is None
    and getattr(options, 'edit', None) is None
):
    ...
```

`getattr(options, 'message', None)` will only be true if the user provided a `-m/--message` option on the command line whereas `cd.message` will be true if there is a non-empty message associated with the commit. For purposes of this predicate, I don't think we want to consider the commit message itself, only whether the user supplied `-m/--message`. Similarly `getattr(options, 'edit', None)` will only be true if the user provided a `--edit` command line option.

In the long run, I'm thinking that `update_commit_data()` should perhaps take explicit arguments for `save_template`, `edit`, and `user_message` instead of inspecting `options` which is wildly different depending on the particular command.

Reply:
> For purposes of this predicate, I don't think we want to consider the commit message itself, only whether the user supplied `-m/--message`.

But that's basically the whole problem with using this for `import`, isn't it?  The general case is that people *don't* want to edit the commit message; rather, they want to use the one that's already there.  The truth table looks like this:

| command | --message | --edit | --save-template | Edit? |
|-|-|-|-|-|
| new / publish | N | n/a | N | Y |
| new / publish | Y | n/a | N | N |
| new / publish | N | n/a | Y | N |
| new / publish | Y | n/a | Y | N |
| import | n/a | N | n/a | N |
| import | n/a | Y | n/a | Y |

This is case where having `update_commit_data()` take an `invoke_editor` boolean might make more sense, and have the options logic decoded by the callers.

If you want I can refactor it that way.  OTOH, I only really wrote this patch so that the second patch would be usable for me.  I'm also happy to have you just take the `--keepcr` patch and fix the `--edit` issue the way you think best.

Reply:
Oh, it should also be pointed out: `cd.message` will be set further up the function if there's a valid `options.message`.  So if `--message` is provided `cd.message` will always be non-`None`.  The only risk in the logic of my submission would be if `cd.message` was somehow set by `publish` or `new` when no `--message` was provided.